crates/containerd-shim-wasmtime/src/http_proxy.rs

```diff
@@ -59,45 +62,65 @@ pub(crate) async fn serve_conn(
     socket.bind(addr)?;
 
     let listener = socket.listen(backlog)?;
+    let tracker = TaskTracker::new();
 
     log::info!("Serving HTTP on http://{}/", listener.local_addr()?);
 
     let env = env.into_iter().collect();
-    let handler = Arc::new(ProxyHandler::new(instance, env));
+    let handler = Arc::new(ProxyHandler::new(instance, env, tracker.clone()));
 
     loop {
-        let (stream, _) = listener.accept().await?;
-        log::debug!("New connection");
-
-        let stream = TokioIo::new(stream);
-        let h = handler.clone();
-
-        tokio::spawn(async {
-            if let Err(e) = http1::Builder::new()
-                .keep_alive(true)
-                .serve_connection(
-                    stream,
-                    hyper::service::service_fn(move |req| h.clone().handle_request(req)),
-                )
-                .await
-            {
-                log::error!("error: {e:?}");
+        tokio::select! {
+            // listen to cancellation requests
+            _ = cancel.cancelled() => {
+                break;
             }
-        });
+            res = listener.accept() => {
+                let (stream, _) = res?;
```

Comment:
**question:** What does it mean for `res` to  be `Err`?
The current behaviour triggers a forced shutdown, is that what we want?
Otherwise we can do something like
```suggestion
                let Ok(stream, _) = res else {
                    log::info!("Error accepting new connection, triggering graceful shutdown.");
                    break;
                };
```

and follow the graceful shutdown path.

Reply:
Good catch! I now changed this and followed `axum` strategy to deal with errors in accept connection.